File: app/hex_math.py
from typing import Tuple, List, Set

# Type alias for clarity
Hex = Tuple[int, int]

# Directions in Axial Coordinates (q, r)
# (1, 0), (1, -1), (0, -1), (-1, 0), (-1, 1), (0, 1)
HEX_DIRECTIONS = [
    (1, 0), (1, -1), (0, -1),
    (-1, 0), (-1, 1), (0, 1)
]

def add_hex(a: Hex, b: Hex) -> Hex:
    return (a[0] + b[0], a[1] + b[1])

def subtract_hex(a: Hex, b: Hex) -> Hex:
    return (a[0] - b[0], a[1] - b[1])

def get_neighbors(hex: Hex) -> List[Hex]:
    """Returns the 6 constituent neighbors of a hex."""
    return [add_hex(hex, d) for d in HEX_DIRECTIONS]

def hex_distance(a: Hex, b: Hex) -> int:
    """Calculates Manhattan distance on a hex grid."""
    vec = subtract_hex(a, b)
    return (abs(vec[0]) + abs(vec[0] + vec[1]) + abs(vec[1])) // 2

def are_neighbors(a: Hex, b: Hex) -> bool:
    return hex_distance(a, b) == 1

def get_common_neighbors(a: Hex, b: Hex) -> List[Hex]:
    """Returns neighbors shared by two hexes."""
    a_neighbors = set(get_neighbors(a))
    b_neighbors = set(get_neighbors(b))
    return list(a_neighbors.intersection(b_neighbors))

def is_connected(hexes: Set[Hex]) -> bool:
    """Checks if a set of hexes forms a single connected component."""
    if not hexes:
        return True
    
    start = next(iter(hexes))
    visited = {start}
    queue = [start]
    
    while queue:
        current = queue.pop(0)
        for neighbor in get_neighbors(current):
            if neighbor in hexes and neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
                
    return len(visited) == len(hexes)

File: cpp/src/hex_math.cpp
#include "hex_math.hpp"
#include <algorithm>
#include <queue>
#include <cmath>
#include <sstream>

namespace bugs {

Hex add_hex(const Hex& a, const Hex& b) {
    return {a.first + b.first, a.second + b.second};
}

Hex subtract_hex(const Hex& a, const Hex& b) {
    return {a.first - b.first, a.second - b.second};
}

std::vector<Hex> get_neighbors(const Hex& hex) {
    std::vector<Hex> neighbors;
    neighbors.reserve(6);
    for (const auto& dir : HEX_DIRECTIONS) {
        neighbors.push_back(add_hex(hex, dir));
    }
    return neighbors;
}

int hex_distance(const Hex& a, const Hex& b) {
    auto vec = subtract_hex(a, b);
    return (std::abs(vec.first) + std::abs(vec.first + vec.second) + std::abs(vec.second)) / 2;
}

bool are_neighbors(const Hex& a, const Hex& b) {
    return hex_distance(a, b) == 1;
}

std::vector<Hex> get_common_neighbors(const Hex& a, const Hex& b) {
    auto a_neighbors = get_neighbors(a);
    auto b_neighbors = get_neighbors(b);
    
    std::unordered_set<Hex, HexHash> a_set(a_neighbors.begin(), a_neighbors.end());
    std::vector<Hex> common;
    
    for (const auto& n : b_neighbors) {
        if (a_set.count(n)) {
            common.push_back(n);
        }
    }
    
    return common;
}

bool is_connected(const std::unordered_set<Hex, HexHash>& hexes) {
    if (hexes.empty()) {
        return true;
    }
    
    // BFS from arbitrary starting hex
    Hex start = *hexes.begin();
    std::unordered_set<Hex, HexHash> visited;
    std::queue<Hex> queue;
    
    visited.insert(start);
    queue.push(start);
    
    while (!queue.empty()) {
        Hex current = queue.front();
        queue.pop();
        
        for (const Hex& neighbor : get_neighbors(current)) {
            if (hexes.count(neighbor) && !visited.count(neighbor)) {
                visited.insert(neighbor);
                queue.push(neighbor);
            }
        }
    }
    
    return visited.size() == hexes.size();
}

std::string coord_to_key(const Hex& h) {
    return std::to_string(h.first) + "," + std::to_string(h.second);
}

Hex key_to_coord(const std::string& key) {
    std::istringstream iss(key);
    int q, r;
    char comma;
    iss >> q >> comma >> r;
    return {q, r};
}

} // namespace bugs
